package com.gnoemes.shikimori.di.character;

import moxy.MvpPresenter;
import com.gnoemes.shikimori.data.repository.roles.CharacterRepository;
import com.gnoemes.shikimori.data.repository.roles.RestCharacterRepositoryImpl;

import com.gnoemes.shikimori.di.base.modules.BaseChildFragmentModule;
import com.gnoemes.shikimori.di.base.scopes.BottomChildScope;
import com.gnoemes.shikimori.domain.roles.CharacterInteractor;
import com.gnoemes.shikimori.domain.roles.CharacterInteractorImpl;
import com.gnoemes.shikimori.presentation.presenter.character.CharacterPresenter;
import com.gnoemes.shikimori.presentation.presenter.character.converter.CharacterDetailsViewModelConverter;
import com.gnoemes.shikimori.presentation.presenter.character.converter.CharacterDetailsViewModelConverterImpl;
import com.gnoemes.shikimori.presentation.presenter.common.converter.DetailsContentViewModelConverter;
import com.gnoemes.shikimori.presentation.presenter.common.converter.DetailsContentViewModelConverterImpl;
import com.gnoemes.shikimori.presentation.view.character.CharacterFragment;

import javax.inject.Named;

import androidx.fragment.app.Fragment;
import dagger.Binds;
import dagger.Module;
import dagger.Reusable;

@Module(includes = {
        BaseChildFragmentModule.class,
})
public interface CharacterModule {

    @Binds
    CharacterInteractor bindCharacterInteractor(CharacterInteractorImpl interactor);

    @Binds
    CharacterRepository bindCharacterRepository(RestCharacterRepositoryImpl repository);
    @Binds
    CharacterDetailsViewModelConverter bindCharacterDetailsViewModelConterter(CharacterDetailsViewModelConverterImpl converter);

    @Binds
    @Reusable
    DetailsContentViewModelConverter bindDetailsContentViewModelConverter(DetailsContentViewModelConverterImpl converter);

    @Binds
    MvpPresenter bindPresenter(CharacterPresenter presenter);

    @Binds
    @Named(BaseChildFragmentModule.CHILD_FRAGMENT)
    @BottomChildScope
    Fragment bindFragment(CharacterFragment fragment);
}
